src/sparse/merkle_tree.rs

```diff
@@ -166,40 +162,49 @@ where
             self.storage.remove(&node.hash())?;
         }
 
-        let mut non_placeholder_reached = false;
+        let path = requested_leaf_node.leaf_key();
+        let mut side_nodes_iter = side_nodes.iter();
+
+        // The deleted leaf is replaced by a placeholder. Build the tree upwards starting with the
+        // placeholder.
         let mut current_node = Node::create_placeholder();
-        let n = side_nodes.len();
-        for i in 0..n {
-            let side_node = &side_nodes[i];
-            if current_node.is_placeholder() {
-                if side_node.is_leaf() {
-                    current_node = side_node.clone();
-                    continue;
-                } else {
-                    non_placeholder_reached = true;
-                }
-            }
 
-            if !non_placeholder_reached && side_node.is_placeholder() {
-                continue;
-            } else if !non_placeholder_reached {
-                non_placeholder_reached = true;
+        // If the first side node is a leaf, it means the ancestor node is now parent to a
+        // placeholder (the deleted leaf node) and a leaf node (the first side node). We can
+        // immediately discard the ancestor node from further calculation and attach the orphaned
+        // leaf node to its next ancestor. Any subsequent ancestor nodes composed of this leaf node
+        // and a placeholder must be similarly discarded from further calculation. We then create a
+        // valid ancestor node for the orphaned leaf node by joining it with the earliest
+        // non-placeholder side node.
+        let first_side_node = side_nodes.first();
+        if first_side_node.is_some() && first_side_node.unwrap().is_leaf() {
+            side_nodes_iter.next();
+            current_node = first_side_node.unwrap().clone();
+
+            // Advance the side node iterator to the next non-placeholder node. This may be either
+            // another leaf node or an internal node.
+            // If only placeholder nodes exist beyond the first leaf node, then that leaf node is,
+            // in fact, the new root node.
+            // Using `find(..)` advances the iterator beyond the next non-placeholder side node and
```

Comment:
Is there a way to link to the `find` function in this Rust comment? Or is that only for docstrings.

Reply:
I do believe that should be possible! I will update the comment with an inline link.

Reply:
I do believe that should be possible! I will update the comment with an inline link.

Reply:
Welp... turns out it's not possible to add a link in the same way that you can do that for docstrings. 
What I can do is simply add a note to check out the url at the end of the comment. Something like:
```
// ...
// See https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find
```